### PR TITLE
make client side know container removal in progress

### LINKED ISSUE #26073: If a container removal is in progress, remove it again seems success?
Hi, All,

I find that when a container is in removal progress, if a user executes a `docker rm` against the container again, docker cli will print the container name and potentially mean removal successfully.

I think this is somewhat unreasonable, since client side **NEVER** has idea about removalInProgress. Maybe we could throw an error and transfer it to the client side.

Here is the server side code, https://github.com/docker/docker/blob/master/daemon/delete.go#L27-L30
```
// Container state RemovalInProgress should be used to avoid races.
	if inProgress := container.SetRemovalInProgress(); inProgress {
		return nil
	}
```

Just a thought for client side. :)

### PR DESCRIPTION
**Pull request by allencloud**
Fixes #26073 
make client side know container removal in progress

Signed-off-by: allencloud allen.sun@daocloud.io

### CODE REVIEW

File: daemon/delete.go [hunk 1]
```diff
@@ -26,7 +26,7 @@ func (daemon *Daemon) ContainerRm(name string, config *types.ContainerRmConfig)
 
 	// Container state RemovalInProgress should be used to avoid races.
 	if inProgress := container.SetRemovalInProgress(); inProgress {
-		return nil
+		return fmt.Errorf("container %s removal is in progress by other client", name)
```
Comment (cpuguy83) [inline daemon/delete.go]: How about `"container %s is already being removed"`

Comment (cpuguy83) [inline daemon/delete.go]: maybe better... `"removal of container %s is already in progress"`

Comment (allencloud) [inline daemon/delete.go]: @cpuguy83 Oh, I think the "removal of ..." is better. I will update this PR. Thanks.

File: daemon/delete_test.go [hunk 2]
```diff
@@ -34,9 +35,9 @@ func TestContainerDoubleDelete(t *testing.T) {
...
 	container.SetRemovalInProgress()
 
-	// Try to remove the container when its start is removalInProgress.
-	// It should ignore the container and not return an error.
-	if err := daemon.ContainerRm(container.ID, &types.ContainerRmConfig{ForceRemove: true}); err != nil {
-		t.Fatal(err)
+	// Try to remove the container when its state is removalInProgress.
+	// It should return an error indicating it is under removal progress.
+	if err := daemon.ContainerRm(container.ID, &types.ContainerRmConfig{ForceRemove: true}); err == nil {
+		t.Fatalf("expected err: %v, got nil", fmt.Errorf("container %s removal is in progress by other client", container.ID))
```
Comment (mlaventure) [inline daemon/delete_test.go]: The expected part doesn't match the actual implementation (which I actually prefer :)). Could you replace it with what you put in `daemon/delete.go`?

Thanks

Comment (allencloud) [inline daemon/delete_test.go]: sorry for my missing. Thanks a lot @mlaventure PTAL

File: daemon/delete_test.go [hunk 3]
```diff
@@ -34,9 +35,9 @@ func TestContainerDoubleDelete(t *testing.T) {
...
 	container.SetRemovalInProgress()
 
-	// Try to remove the container when its start is removalInProgress.
-	// It should ignore the container and not return an error.
-	if err := daemon.ContainerRm(container.ID, &types.ContainerRmConfig{ForceRemove: true}); err != nil {
-		t.Fatal(err)
+	// Try to remove the container when its state is removalInProgress.
+	// It should return an error indicating it is under removal progress.
+	if err := daemon.ContainerRm(container.ID, &types.ContainerRmConfig{ForceRemove: true}); err == nil {
+		t.Fatalf("expected err: %v, got nil", "removal of container %s is already in progress", container.ID)
```
Comment (mlaventure) [inline daemon/delete_test.go]: You've mistakenly removed the needed `fmt.Sprintf()`

Comment (allencloud) [inline daemon/delete_test.go]: sorry for my mistake. Thanks for you help. :) updated that. Actually so sleepy that I have to go to bed.  See you tomorrow, my friend.

**Comment by mlaventure** [comment]
LGTM.

**Comment by allencloud** [comment]
ping @cpuguy83 @tiborvass, PTAL, THANKS.

**Comment by allencloud** [comment]
@cpuguy83 @tiborvass PTAL

**Comment by thaJeztah** [comment]
"SGTM", but needs changes to the documentation;

- A mention in the API changes: https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v125-api-changes
- The status codes returned by the endpoint (probably); https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.25.md#remove-a-container
- If this change is not versioned (i.e., it also affects older versions of the API), older API docs would need to have the same update.

@bfirsh @dnephin would this be a change that needs to be versioned?

**Comment by dnephin** [comment]
I don't think it needs to be versioned, it's a change in behaviour, but the API looks the same.

**Comment by thaJeztah** [comment]
alright, LGTM, moving to docs review

**Comment by allencloud** [comment]
@mlaventure @thaJeztah @dnephin @cpuguy83 
I am still wondering that if this case should return a status code of 409?

**Comment by cpuguy83** [comment]
400 works for me.

**Comment by mlaventure** [comment]
Either code works for me

**Comment by allencloud** [comment]
Thanks for all your feedback.
Now I added two kinds of change in this PR:
1. return an error with status code 400;
2. remove error judge of 
```
if strings.Contains(strings.ToLower(err.Error()), "prefix can't be empty") {		
 -			return fmt.Errorf("no such container: \"\"")		
 -		}
```
As in func `getContainer`, there is empty container name judging logic there.
```
func (daemon *Daemon) GetContainer(prefixOrName string) (*container.Container, error) {
	if len(prefixOrName) == 0 {
		return nil, errors.NewBadRequestError(fmt.Errorf("No container name or ID supplied"))
	}
        ...
}
```
So the code is redundant in `container_routes.go`.

PTAL @mlaventure @thaJeztah @dnephin @cpuguy83 @tiborvass

File: api/server/router/container/container_routes.go [hunk 4]
```diff
@@ -386,10 +385,6 @@ func (s *containerRouter) deleteContainers(ctx context.Context, w http.ResponseW
 	}
 
 	if err := s.backend.ContainerRm(name, config); err != nil {
-		// Force a 404 for the empty string
```
Comment (mlaventure) [inline api/server/router/container/container_routes.go]: Why was this removed?

Comment (allencloud) [inline api/server/router/container/container_routes.go]: Since if name is empty, in the implementation of `ContainerRm`, `getContainer` will take into empty string into consideration, and return an error 
```
func (daemon *Daemon) GetContainer(prefixOrName string) (*container.Container, error) {
	if len(prefixOrName) == 0 {
		return nil, errors.NewBadRequestError(fmt.Errorf("No container name or ID supplied"))
	}
        ....
}
```

So if the string is empty, will return an error of `fmt.Errorf("No container name or ID supplied")` instead of `ErrEmptyPrefix = errors.New("Prefix can't be empty")`. So remove the code above.

Comment (allencloud) [inline api/server/router/container/container_routes.go]: If it is not reasonable, I can add it back. Open for input.

Comment (mlaventure) [inline api/server/router/container/container_routes.go]: Just realized this code would produce a 500 error and not a 404 anyhow. I guess it's fine to remove then. It'll return a 400 error instead.

**Review by mlaventure — APPROVED** [review]
LGTM

**Review by dnephin — APPROVED** [review]
LGTM

**Comment by vdemeester** [comment]
Moving to docs review /cc @thaJeztah

**Comment by allencloud** [comment]
ping @thaJeztah

**Comment by thaJeztah** [comment]
@allencloud I left some suggestions earlier; https://github.com/docker/docker/pull/26074#issuecomment-248904533

**Comment by allencloud** [comment]
@thaJeztah 
Sorry for my missing that. Added, PTAL. 😄 
1. this change is added in docker_remote_api.md#v125-api-changes
2. no change in status code;
3. no need to be versioned as discussed above.

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by thaJeztah** [comment]
ping @vdemeester for review/merge

**Review by vdemeester — APPROVED** [review]
LGTM 🐸
